src/firmware/types.rs

```diff
@@ -202,3 +210,512 @@ pub struct SnpPlatformStatus {
     /// Reported TCB version.
     pub reported_tcb_version: TcbVersion,
 }
+
+#[repr(C)]
+pub struct SnpConfig {
+    reserved: [u8; 52],
+}
+
+impl Default for SnpConfig {
+    fn default() -> Self {
+        Self { reserved: [0; 52] }
+    }
+}
+
+#[repr(C)]
+pub struct SnpSetExtConfig<'a> {
+    /// Address of the SnpConfig or 0 when reported_tcb does not need
+    /// to be updated.
+    pub config_address: Option<&'a SnpConfig>,
+
+    /// Address of extended guest request certificate chain or 0 when
+    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    pub certs_address: Option<*mut c_void>,
+
+    /// Length of the certificates.
+    pub certs_len: u32,
+}
+
+#[repr(C)]
+pub struct SnpGetExtConfig<'a> {
+    /// Address of the SnpConfig or 0 when reported_tcb does not need
+    /// to be updated.
+    pub config_address: Option<&'a SnpConfig>,
+
+    /// Address of extended guest request certificate chain or 0 when
+    /// previous certificate should be removed on SNP_SET_EXT_CONFIG.
+    pub certs_address: Option<*mut c_void>,
+
+    /// Length of the certificates.
+    pub certs_len: u32,
+}
+
+/// Locates the non-zero based index bit field of a specified number.
+///
+/// `$field`: expr - The number in question the bit should be taken from.
+/// `$bit`: expr - The one-based index bit being requested.
+/// `$dt`: ty - The data type of the field (for alignment reasons).
+///
+/// # Examples:
+///
+/// To find the 3rd bit field of Guest Field Select of a Derived Key Request:
+///
+/// ```
+/// let mut key_req: DerivedKeyReq = Default::default();
+/// key_req.guest_field_select = 4; /// 0b0..100 the third bit should be on.
+/// let res = bit_val\!(key_req.guest_field_select, 3, u64);
+/// assert_eq!(res, 1);
+/// ```
+///
+#[macro_export]
+macro_rules! bit_val {
+    ($field: expr, $bit: expr, $dt: ty) => {
+        {
+            ($field & (1 as $dt << ($bit as $dt - 1 as $dt))) >> ($bit - 1 as $dt)
+        }
+    }
+}
+
+#[cfg(feature = "openssl")]
+fn swap_endian(bytes: &mut [u8]) -> Vec<u8> {
+    let mut retval: Vec<u8> = Vec::new();
+
+    for byte in bytes.iter().rev() {
+        retval.push(*byte);
+    }
+    retval
+}
+
+/// Information provided by the guest owner for requesting an attestation
+/// report from the AMD Secure Processor.
+#[repr(C)]
+pub struct ReportReq {
+    /// Guest-provided data to be included into the attestation report
+    pub report_data: [u8; 64],
+
+    /// The VMPL to put into the attestation report. Must be greater than or
+    /// equal to the current VMPL and at most three.
+    pub vmpl: u32,
+
+    /// Reserved memory slot, but be zero.
+    reserved: [u8; 28],
+}
+
+impl Default for ReportReq {
+    fn default() -> Self {
+        Self {
+            report_data: [0x0F; 64],
```

Comment:
Why is report_data filled with `0x0f` bytes? Zero-filled sounds more like a default value.

Reply:
Whoops, yep this can be zeros. This is just carry over.